[quant][observer] Add per channel observer

hx89 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#25887 [quant][observer] Add per channel observer**

Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

Commit:
[quant][observer] Add per channel observer



Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

Commit:
Update on "[quant][observer] Add per channel observer"


Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -775,6 +775,43 @@ def test_minmax_observer(self, qdtype, qscheme, reduce_range):
+           qscheme=st.sampled_from((torch.per_channel_affine, torch.per_channel_symmetric)))
+    def test_per_channel_minmax_observer(self, qdtype, qscheme):
+        myobs = PerChannelMinMaxObserver(ch_axis=1, dtype=qdtype, qscheme=qscheme)
+        x = torch.tensor([[1.0, 2.0, 2.0, 3.0, 4.0, 6.0], [-4.0, 5.0, 5.0, 6.0, 7.0, 8.0]])
```
Please add a test for 4D tensors, I found a bug in the original implementation as it worked only for 2d

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -137,7 +186,48 @@ def calculate_qparams(self):
+        self.max_vals = None
+
+    def forward(self, x):
+        min_vals = self.min_vals
```
Please change to (the current code doesnt work for 4d tensors):
``` python
    def forward(self, x):
        min_vals = self.min_vals
        max_vals = self.max_vals
        x_dim = x.size()

        y = torch.reshape(x, (x_dim[self.ch_axis] , int(torch.numel(x) / x_dim[self.ch_axis]))).detach()
        if min_vals is None or max_vals is None:
            min_vals = torch.min(y, 1)[0]
            max_vals = torch.max(y, 1)[0]
        else:
            min_vals = torch.min(torch.min(y, 1)[0], min_vals)
            max_vals = torch.max(torch.max(y, 1)[0], max_vals)
        self.min_vals = min_vals.detach()
        self.max_vals = max_vals.detach()
        return x
```

raghuramank100 reviewed (changes requested):
Please see comments

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

hx89 commented:
Addressed comments.

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -775,6 +775,45 @@ def test_minmax_observer(self, qdtype, qscheme, reduce_range):
+        buf.seek(0)
+        loaded = torch.jit.load(buf)
+        self.assertEqual(obs.calculate_qparams(), loaded.calculate_qparams())
+
```
This is a general request: We need to test that all observers are serializable. We currently only test for scriptability. This is needed for quantization aware training support. Can you add support in a seperate PR?
https://github.com/pytorch/pytorch/blob/9f6b6b81010e6f2edbf5faeb521284b1276ee1de/test/test_quantized_nn_mods.py#L254 to https://github.com/pytorch/pytorch/blob/9f6b6b81010e6f2edbf5faeb521284b1276ee1de/test/test_quantized_nn_mods.py#L303 provides a refefrence.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -45,15 +53,47 @@ def forward(self, x):
+            qparam = self._calculate_qparams(
+                min_vals[i], max_vals[i]
+            )
+            scales[i] = float(qparam[0])
```
Do you need to convert to float/int here? Both qparam and scales are tensors

hx89 replied:
Without float/int, scriptable test would have error:
> RuntimeError: output with shape [] doesn't match the broadcast shape [1]
probably because scales[i] is the scalar.

raghuramank100 commented on torch/quantization/observer.py:
```diff
@@ -137,7 +186,51 @@ def calculate_qparams(self):
+        self.max_vals = None
+
+    def forward(self, x):
+        min_vals = self.min_vals
```
To ensure that autograd does not see this part during training:, please do:
```python
def forward(self, x):
    with torch.no_grad():
 <Body of code> 
```

hx89 replied:
Seems after the change scriptable test would fail with the error:
> torch.jit.frontend.UnsupportedNodeError: with statements aren't supported:

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -775,6 +775,45 @@ def test_minmax_observer(self, qdtype, qscheme, reduce_range):
 
+    @given(qdtype=st.sampled_from((torch.qint8, torch.quint8)),
+           qscheme=st.sampled_from((torch.per_channel_affine, torch.per_channel_symmetric)))
+    def test_per_channel_minmax_observer(self, qdtype, qscheme):
```
Please sweep over axis and reduce_range for the tests.

raghuramank100 reviewed (changes requested):
Few more comments.

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

raghuramank100 commented on test/test_quantization.py:
```diff
@@ -775,6 +775,64 @@ def test_minmax_observer(self, qdtype, qscheme, reduce_range):
+        result = myobs(x)
+        self.assertEqual(result, x)
+        qparams = myobs.calculate_qparams()
+        if ch_axis == 0:
```
For better readability, please modify this to reuse code, example would look like this:
```python
          self.assertEqual(myobs.min_vals, ref_min_vals[ ch_axis])
            self.assertEqual(myobs.max_vals, ref_max_vals[ch_axis])
            if qscheme == torch.per_channel_symmetric:
                ref_scales = per_channel_symmetric_ref_scales[ch_axis]
                ref_zero_points = [0, 0] if qdtype is torch.qint8 else [128, 128]
            else:
                ref_scales = per_channel_affine_ref_scales[ch_axis]
                ref_zero_points = per_channel_affine_qint8_zp[ch_axis] if qdtype is torch.qint8 else  per_channel_affine_quint8_zp[ch_axis]
```

raghuramank100 reviewed (approved):
Approved,  please see comments to address prior to submitting

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]

Commit:
Update on "[quant][observer] Add per channel observer"


Add per channel observer to compute the qparams for each channel.

Differential Revision: [D17137226](https://our.internmc.facebook.com/intern/diff/D17137226/)

[ghstack-poisoned]
